File: Plot_Tools/Plot_SPL_DiffFreqRange.py
import numpy as np
import soundfile as sf
from scipy.signal import resample, welch
from scipy.fft import fft, ifft
import matplotlib.pyplot as plt
from scipy.signal import iirfilter, lfilter
import scipy.signal as signal

def design_a_weighting(fs):
    """
    * This function is used to design an a-weighting filter according to the sampling rate
    @ params:
    - fs: sampling rate [float]
    @ returns:
    - b: Numerator polynomials of the IIR [ndarray]
    - a: Denominator polynomials of the IIR [ndarray]
    """
    f1 = 20.598997
    f2 = 107.65265
    f3 = 737.86223
    f4 = 12194.217
    A1000 = 1.9997

    numerators = [(2*np.pi * f4)**2 * (10**(A1000/20)), 0, 0, 0, 0]
    denominators = np.convolve([1, 4*np.pi * f4, (2*np.pi * f4)**2],
                               [1, 4*np.pi * f1, (2*np.pi * f1)**2])
    denominators = np.convolve(np.convolve(denominators, 
                                           [1, 2*np.pi * f3]),
                               [1, 2*np.pi * f2])
    
    b, a = signal.bilinear(numerators, denominators, fs)
    return b, a

def a_weighting(waveform, fs):
    b, a = design_a_weighting(fs)
    filtered_wave = signal.lfilter(b,a,waveform)
    return filtered_wave

def bandpower(signal, fs, fmin, fmax, freq_resoluton):
    """
    Compute band power within a specified frequency range using Welch's method.
    
    Parameters:
    - signal: input signal [ndarray]
    - fs: sampling rate [float]
    - fmin: minimum frequency of interest [float]
    - fmax: maximum frequency of interest [float]
    
    Returns:
    - band_power: computed band power [float]
    """
    nperseg = int(fs / freq_resoluton)
    f, Pxx = welch(signal, fs, nperseg=nperseg,nfft=fs, noverlap=nperseg*0.75)
    ind_min = np.argmax(f > fmin) - 1
    ind_max = np.argmax(f > fmax) - 1
    return np.trapz(Pxx[ind_min: ind_max], f[ind_min: ind_max])

def calculate_psd(signal, fs, freq_start, freq_end, freq_resolution):
    """
    Calculate the Power Spectral Density (PSD) of the signal and convert it to dB SPL.

    Parameters:
    - signal: Input signal
    - fs: Sampling frequency
    - freq_start: Start frequency for analysis
    - freq_end: End frequency for analysis
    - freq_resolution: Desired frequency resolution

    Returns:
    - f: Frequency array
    - Pxx_dB: PSD in dB SPL
    """
    nperseg = int(fs / freq_resolution)  # Number of samples per segment
    f, Pxx = welch(signal, fs, nperseg=nperseg, nfft=fs, noverlap=nperseg*0.75)

    # Select the frequency range
    mask = (f >= freq_start) & (f <= freq_end)
    f = f[mask]
    Pxx = Pxx[mask]
    return f, Pxx

def plot_rmsspl(signal, fs, freq_start, freq_end, label):
    """
    Plot RMS SPL vs. frequency for a given signal.
    
    Parameters:
    - signal: input signal [ndarray]
    - fs: sampling rate [float]
    - freq_start: starting frequency for plot [float]
    - freq_end: ending frequency for plot [float]
    - label: label for the signal [str]
    
    Returns:
    - SPLfft: RMS SPL value [float]
    """
    # f, Pxx = welch(signal, fs, nperseg=1024)
    f, Pxx = calculate_psd(signal,fs, freq_start, freq_end, 5)
    Pxx = 10 * np.log10(Pxx / (20e-6)**2)
    band_power = bandpower(signal, fs, freq_start, freq_end, 5)
    SPLfft = 10 * np.log10(band_power / (20e-6)**2)
    
    plt.plot(f, Pxx, label=f'{label}: {SPLfft:.2f} dBA')
    plt.xlim([freq_start, freq_end])
    plt.xlabel('Frequency (Hz)')
    plt.ylabel('SPL (dBA)')
    plt.grid(True)
    
    return SPLfft

# Read the data
mic_data1, fs1 = sf.read("Real Noise/Aircraft.wav")
mic_data2, fs2 = sf.read("Real Noise/Mix_Aircraft_Traffic.wav")

# Resample if sampling rates are different
if fs1 != fs2:
    if fs1 > fs2:
        mic_data2 = resample(mic_data2, int(len(mic_data2) * fs1 / fs2))
        fs2 = fs1
    else:
        mic_data1 = resample(mic_data1, int(len(mic_data1) * fs2 / fs1))
        fs1 = fs2

# Pad the shorter signal with zeros if lengths are different
if len(mic_data1) > len(mic_data2):
    mic_data2 = np.pad(mic_data2, (0, len(mic_data1) - len(mic_data2)), 'constant')
else:
    mic_data1 = np.pad(mic_data1, (0, len(mic_data2) - len(mic_data1)), 'constant')

# Apply A-weighting
mic_data1_a = a_weighting(mic_data1, fs1)
mic_data2_a = a_weighting(mic_data2, fs2)

# Plot the corresponding SPL vs. frequency
plt.figure(figsize=(12, 10))

# Full frequency range
plt.subplot(221)
plot_rmsspl(mic_data1_a, fs1, 50, 8000, 'Mic1')
plot_rmsspl(mic_data2_a, fs2, 50, 8000, 'Mic2')
plt.title('SPL vs. Frequency in 50Hz-8000Hz')
plt.legend()

# Low frequency range
plt.subplot(222)
plot_rmsspl(mic_data1_a, fs1, 50, 200, 'Mic1')
plot_rmsspl(mic_data2_a, fs2, 50, 200, 'Mic2')
plt.title('SPL vs. Frequency in 50Hz-200Hz')
plt.legend()

# Middle frequency range
plt.subplot(223)
plot_rmsspl(mic_data1_a, fs1, 200, 2000, 'Mic1')
plot_rmsspl(mic_data2_a, fs2, 200, 2000, 'Mic2')
plt.title('SPL vs. Frequency in 200Hz-2000Hz')
plt.legend()

# High frequency range
plt.subplot(224)
plot_rmsspl(mic_data1_a, fs1, 2000, 8000, 'Mic1')
plot_rmsspl(mic_data2_a, fs2, 2000, 8000, 'Mic2')
plt.title('SPL vs. Frequency in 2000Hz-8000Hz')
plt.legend()

plt.tight_layout()
plt.show()

File: Plot_Tools/Plot_SPL_DiffFreqRange.m
clear 
close all
clc

addpath 'Real Noise'

% Set the low frequency range
low_freq_start = 50;
low_freq_end = 200;
% Set the middle frequency range
mid_freq_start = 200;
mid_freq_end = 2000;
% Set the high frequency range
high_freq_start = 2000;
high_freq_end = 8000;
% Set the full frequency range
full_freq_start = 50;
full_freq_end = 8000;

%% Read the data
[Mic_data1, fs1] = audioread("Aircraft.wav");
[Mic_data2, fs2] = audioread("Mix_Aircraft_Traffic.wav");

% Resample if sampling rates are different
if fs1 ~= fs2
    if fs1 > fs2
        Mic_data2 = resample(Mic_data2, fs1, fs2);
        fs2 = fs1;
    else
        Mic_data1 = resample(Mic_data1, fs2, fs1);
        fs1 = fs2;
    end
end

% Pad the shorter signal with zeros if lengths are different
if length(Mic_data1) > length(Mic_data2)
    Mic_data2 = [Mic_data2; zeros(length(Mic_data1) - length(Mic_data2), 1)];
else
    Mic_data1 = [Mic_data1; zeros(length(Mic_data2) - length(Mic_data1), 1)];
end

weighting = weightingFilter('A-weighting', fs1);
Mic_data1_A = weighting(Mic_data1);
Mic_data2_A = weighting(Mic_data2);

%% Plot the corresponding SPL vs. frequency
figure();
% full frequency range
subplot(221)
SPLfft_1 = plot_RMSSPL(Mic_data1_A, fs1, full_freq_start, full_freq_end, 'Mic1');
hold on
SPLfft_2 = plot_RMSSPL(Mic_data2_A, fs2, full_freq_start, full_freq_end, 'Mic2');
legend_1 = ['Mic1 ', num2str(SPLfft_1), ' dBA'];
legend_2 = ['Mic2 ', num2str(SPLfft_2), ' dBA'];
title(['SPL vs. Frequency in ', num2str(full_freq_start), 'Hz-', num2str(full_freq_end), 'Hz']);
xlim([full_freq_start, full_freq_end]);
legend(legend_1, legend_2);
% low frequency range
subplot(222)
SPLfft_1 = plot_RMSSPL(Mic_data1_A, fs1, low_freq_start, low_freq_end, 'Mic1');
hold on
SPLfft_2 = plot_RMSSPL(Mic_data2_A, fs2, low_freq_start, low_freq_end, 'Mic2');
legend_1 = ['Mic1 ', num2str(SPLfft_1), ' dBA'];
legend_2 = ['Mic2 ', num2str(SPLfft_2), ' dBA'];
title(['SPL vs. Frequency in ', num2str(low_freq_start), 'Hz-', num2str(low_freq_end), 'Hz']);
xlim([low_freq_start, low_freq_end]);
legend(legend_1, legend_2);
% middle frequency range
subplot(223)
SPLfft_1 = plot_RMSSPL(Mic_data1_A, fs1, mid_freq_start, mid_freq_end, 'Mic1');
hold on
SPLfft_2 = plot_RMSSPL(Mic_data2_A, fs2, mid_freq_start, mid_freq_end, 'Mic2');
legend_1 = ['Mic1 ', num2str(SPLfft_1), ' dBA'];
legend_2 = ['Mic2 ', num2str(SPLfft_2), ' dBA'];
title(['SPL vs. Frequency in ', num2str(mid_freq_start), 'Hz-', num2str(mid_freq_end), 'Hz']);
xlim([mid_freq_start, mid_freq_end]);
legend(legend_1, legend_2);
% high frequency range
subplot(224)
SPLfft_1 = plot_RMSSPL(Mic_data1_A, fs1, high_freq_start, high_freq_end, 'Mic1');
hold on
SPLfft_2 = plot_RMSSPL(Mic_data2_A, fs2, high_freq_start, high_freq_end, 'Mic2');
legend_1 = ['Mic1 ', num2str(SPLfft_1), ' dBA'];
legend_2 = ['Mic2 ', num2str(SPLfft_2), ' dBA'];
title(['SPL vs. Frequency in ', num2str(high_freq_start), 'Hz-', num2str(high_freq_end), 'Hz']);
xlim([high_freq_start, high_freq_end]);
legend(legend_1, legend_2);

%% Function
function [SPLfft] = plot_RMSSPL(signal_A, fs, freq_start, freq_end, mic_label)
    % Compute the power spectral density
    [p, f] = pspectrum(signal_A, fs, 'FrequencyLimits', [freq_start, freq_end], 'FrequencyResolution', 5);
    % Convert to dB SPL
    y = 10*log10(p / (20e-6)^2);
    
    % Plot the spectrum
    if strcmp(mic_label, 'Mic1')    
        plot(f, y, 'linewidth', 1.5, 'Color', [0.8500, 0.3250, 0.0980]);
    else
        plot(f, y, 'linewidth', 1.5, 'Color', [0, 0.4470, 0.7410]);
    end
    
    % Calculate the RMS SPL in the specified frequency range
    band_power = bandpower(signal_A, fs, [freq_start freq_end]);
    SPLfft = 10*log10(band_power / (20e-6)^2);
    SPLfft = roundn(SPLfft, -2);
    
    xlabel("Frequency (Hz)");
    ylabel("SPL (dBA)");
    set(gca, 'fontsize', 18);
    grid on; 
    grid minor;
    hold on;
end